File: 3496-minimum-number-of-seconds-to-make-mountain-height-zero/minimum-number-of-seconds-to-make-mountain-height-zero.rb
# @param {Integer} mountain_height
# @param {Integer[]} worker_times
# @return {Integer}
def min_number_of_seconds(mountain_height, worker_times)
    min_worker_time = 1
    max_worker_time = (mountain_height * (mountain_height + 1) * worker_times.max) / 2

    return binary_search_minimum_total_time(min_worker_time, max_worker_time, mountain_height, worker_times)
end

def binary_search_minimum_total_time(min_worker_time, max_worker_time, mountain_height, worker_times)
    min_total_time = 0

    while min_worker_time <= max_worker_time
        current_time = min_worker_time + (max_worker_time - min_worker_time) / 2
        height_covered_in_current_time = 0

        worker_times.each do |worker_time|
            height_covered_by_worker = ((-1.0 + Math.sqrt(1.0 + (8.0 * current_time) / worker_time)) / 2.0).floor
            height_covered_in_current_time += height_covered_by_worker
        end

        if height_covered_in_current_time < mountain_height
            min_worker_time = current_time + 1
        else
            min_total_time = current_time
            max_worker_time = current_time - 1
        end
    end

    return min_total_time
end

File: 3496-minimum-number-of-seconds-to-make-mountain-height-zero/minimum-number-of-seconds-to-make-mountain-height-zero.js
/**
 * @param {number} mountainHeight
 * @param {number[]} workerTimes
 * @return {number}
 */
var minNumberOfSeconds = function(mountainHeight, workerTimes) {
    let minWorkerTime = 1;
    let maxWorkerTime = mountainHeight * (mountainHeight + 1) / 2 * Math.max(...workerTimes);

    return binarySearchMinimumTotalTime(minWorkerTime, maxWorkerTime, mountainHeight, workerTimes);
};

function binarySearchMinimumTotalTime(minWorkerTime, maxWorkerTime, mountainHeight, workerTimes) {
    let minTotalTime = 0;

    while (minWorkerTime <= maxWorkerTime) {
        let currentTime = Math.floor(minWorkerTime + (maxWorkerTime - minWorkerTime) / 2);
        let heightCoveredInCurrentTime = 0;

        for (const workerTime of workerTimes) {
            let heightCoveredByWorker = Math.floor((-1 + Math.sqrt(1 + (8 * currentTime) / workerTime)) / 2);
            heightCoveredInCurrentTime += heightCoveredByWorker;
        }

        if (heightCoveredInCurrentTime < mountainHeight) {
            minWorkerTime = currentTime + 1;
        } else {
            minTotalTime = currentTime;
            maxWorkerTime = currentTime - 1;
        }
    }

    return minTotalTime;
}
